ledger/participant-integration-api/src/main/scala/platform/partymanagement/PersistentPartyRecordStore.scala

```diff
@@ -0,0 +1,301 @@
+        // Update an existing party record
+        case Some(dbPartyRecord) =>
+          doUpdatePartyRecord(
+            dbPartyRecord = dbPartyRecord,
+            partyRecordUpdate = partyRecordUpdate,
+          )(connection)
+          withPartyRecord(party) { updatedDbPartyRecord =>
+            val annotations = backend.getPartyAnnotations(dbPartyRecord.internalId)(connection)
+            toDomainPartyRecord(
+              payload = updatedDbPartyRecord.payload,
+              annotations = annotations,
+            )
+          }
+        // Party record does not exist
+        case None =>
+          throw PartyRecordNotFoundOnUpdateException
+      }
+    }.map(tapSuccess { updatePartyRecord =>
+      logger.info(s"Updated party record in participant local store: ${updatePartyRecord}")
+    })
+    updateFuture.recoverWith[Result[domain.ParticipantParty.PartyRecord]] {
+      case PartyRecordNotFoundOnUpdateException =>
+        for {
+          // NOTE: We are inspecting a ledger party existence only if we couldn't find the target party record for update.
+          partyExistsOnLedger <- ledgerPartyExists.exists(party)
+          createdPartyRecord <-
+            if (partyExistsOnLedger) {
+              // Optional second transaction:
+              createAndUpdatePartyRecordInTransaction(
+                partyRecordUpdate = partyRecordUpdate,
+                party = party,
+                newPartyRecord = domain.ParticipantParty.PartyRecord(
+                  party = party,
+                  metadata = domain.ObjectMeta.empty,
+                ),
+              )
+            } else {
+              Future.successful(
+                Left(PartyRecordStore.PartyNotFound(party))
+              )
+            }
+        } yield createdPartyRecord
+    }
+  }
+
+  override def getPartyRecord(
+      party: Party
+  )(implicit loggingContext: LoggingContext): Future[Result[ParticipantParty.PartyRecord]] = {
+    inTransaction(_.getPartyRecord) { implicit connection =>
+      withPartyRecord(party) { dbPartyRecord =>
+        val annotations = backend.getPartyAnnotations(dbPartyRecord.internalId)(connection)
+        toDomainPartyRecord(dbPartyRecord.payload, annotations)
+      }
+    }
+  }
+
+  private def doCreatePartyRecord(
+      partyRecord: ParticipantParty.PartyRecord
+  )(connection: Connection): Unit = {
+    val now = epochMicroseconds()
+    val dbParty = PartyRecordStorageBackend.DbPartyRecordPayload(
+      party = partyRecord.party,
+      resourceVersion = 0,
+      createdAt = now,
+    )
+    val internalId = backend.createPartyRecord(
+      partyRecord = dbParty
+    )(connection)
+    partyRecord.metadata.annotations.foreach { case (key, value) =>
+      backend.addPartyAnnotation(
+        internalId = internalId,
+        key = key,
+        value = value,
+        updatedAt = now,
+      )(connection)
+    }
+  }
+
+  private def createAndUpdatePartyRecordInTransaction(
```

Comment:
An argument against doing it, is that it would introduce a new piece of logic: creating a new party record based on updatePartyRecord object - which would have to be kept consistent with a sequence of createNewPartyRecord and updatePartyRecord calls. Current ensures this property by construction. The other approach would necessitate an extra care to have this covered by tests

Reply:
I think it would be the right approach though. There is nothing uncommon in an update falling back to create in case it is a first update and the code required is rather trivial and easy to test.
The benefit of this approach is that at no point in time do we create a record in the database that no-one really asked for

Reply:
Ok